### ISSUE TITLE
Metric update method is called twice in a row

### ISSUE THREAD

**Issue by Ceceu** [PyTorchLightning/pytorch-lightning#4921]
Why method `update` is called twice in a row when a metric is used?

#### Code
For instance, in the [Precision](https://github.com/PyTorchLightning/pytorch-lightning/blob/9801e7694d788c0e57938d66663502f04d13a0d5/pytorch_lightning/metrics/classification/precision_recall.py#L27) metric below
```python
import torch
from torch import nn
from pytorch_lightning.metrics.metric import Metric
from pytorch_lightning.metrics.utils import to_onehot, METRIC_EPS, _input_format_classification_one_hot


class Precision(Metric):

    def __init__(
        self,
        num_classes: int = 1,
        threshold: float = 0.5,
        average: str = 'micro',
        multilabel: bool = False,
        compute_on_step: bool = True,
        dist_sync_on_step: bool = False,
        process_group: Optional[Any] = None,
    ):
        super().__init__(
            compute_on_step=compute_on_step,
            dist_sync_on_step=dist_sync_on_step,
            process_group=process_group,
        )

        self.num_classes = num_classes
        self.threshold = threshold
        self.average = average
        self.multilabel = multilabel

        assert self.average in ('micro', 'macro'), \
            "average passed to the function must be either `micro` or `macro`"

        self.add_state("true_positives", default=torch.zeros(num_classes), dist_reduce_fx="sum")
        self.add_state("predicted_positives", default=torch.zeros(num_classes), dist_reduce_fx="sum")

    def update(self, preds: torch.Tensor, target: torch.Tensor):

        print("call update")
        preds, target = _input_format_classification_one_hot(
            self.num_classes, preds, target, self.threshold, self.multilabel
        )

        # multiply because we are counting (1, 1) pair for true positives
        self.true_positives += torch.sum(preds * target, dim=1)
        self.predicted_positives += torch.sum(preds, dim=1)


    def compute(self):
        if self.average == 'micro':
            return self.true_positives.sum().float() / (self.predicted_positives.sum() + METRIC_EPS)
        elif self.average == 'macro':
            return (self.true_positives.float() / (self.predicted_positives + METRIC_EPS)).mean()
```
when called, it produces the following output:

```python
predictions = torch.rand(2,3)
targets = torch.rand(2,3)

mrr = Precision()
mrr(predictions, targets)
# call update
# call update
# tensor(0.4187)
```

**Comment by SkafteNicki** [comment]
The first call updates the global metric state (the one that is being accumulated over multiple batches) and the second is for calculating the metric value on the current batch. This cannot be combined into one call as we would limit how metric states can be used. 

If you are only interested in only accumulating over multiple batches you can just call `update` and if you are only interested in calculating the metric on the current batch you can use the functional metric interface.

**Comment by SkafteNicki** [comment]
@Ceceu I am closing this, please reopen if more questions on this topic.

**Comment by indigoviolet** [comment]
I ran into this issue just now, and I wonder if this is a limitation of the API: would it be better to have the abstract functions broken out into `compute_state()`, `update_state()` (which receives the results of `compute_state()`), and `reduce_state()` (which returns the synchronized state)? 

Calling `update()` twice under the hood is very unexpected.

**Comment by SkafteNicki** [comment]
@indigoviolet how much of an issue is this? After all the reason why `update` is called twice is due to `forward` both calculating the metric on the current batch and updating the global state. It therefore makes some sense that update is called twice. 
If one only want to update the global state, call `update` directly. If one only want the value for the current batch then calling `update(...).compute()` will do the trick.

**Comment by indigoviolet** [comment]
yes,  but i still feel it is a gotcha which will trip up many new users. At
the very least we should document this behavior.

**Comment by SkafteNicki** [comment]
@indigoviolet I do agree that it is not something explicit written anywhere. Maybe we should make a note in the docs?
You are also right that it is a limitation of the current API, but IMO the simple API is more useful for the general user (even if it has some flaws).

**Comment by indigoviolet** [comment]
yeah I think making a note in the docs is a good step for now.
